src/taproot.rs

```diff
@@ -23,55 +23,36 @@ use std::collections::{BTreeMap, BTreeSet, BinaryHeap};
 use secp256k1_zkp::{self, Secp256k1, Scalar};
 use crate::encode::Encodable;
 
-/// The SHA-256 midstate value for the TapLeaf/elements hash.
-const MIDSTATE_TAPLEAF: [u8; 32] = [
-    185, 165, 93, 95, 240, 236, 34, 5, 36, 232, 194, 245, 220, 195, 78, 19, 172,
-    0, 50, 118, 247, 82, 182, 211, 91, 67, 107, 156, 165, 45, 235, 183
-];
-// (rev) b7eb2da59c6b435bd3b652f7763200ac134ec3dcf5c2e8240522ecf05f5da5b9
-
-/// The SHA-256 midstate value for the TapBranch hash.
-const MIDSTATE_TAPBRANCH: [u8; 32] = [
-    252, 158, 245, 135, 52, 103, 176, 127, 235, 57, 57, 126, 85, 222, 151, 33, 244,
-    104, 173, 199, 58, 32, 119, 252, 160, 87, 213, 147, 185, 136, 180, 140
-];
-// (rev) 8cb488b993d557a0fc77203ac7ad68f42197de557e3939eb7fb0673487f59efc
-
-/// The SHA-256 midstate value for the TapTweak hash.
-const MIDSTATE_TAPTWEAK: [u8; 32] = [
-    7, 183, 63, 121, 138, 46, 7, 245, 251, 66, 173, 40, 201, 174, 109, 157, 27, 32,
-    0, 107, 144, 33, 8, 203, 198, 48, 213, 13, 252, 12, 251, 9
-];
-// (rev) 09fb0cfc0dd530c6cb0821906b00201b9d6daec928ad42fbf5072e8a793fb707
-
-/// The SHA-256 midstate value for the TapSighash hash.
-const MIDSTATE_TAPSIGHASH: [u8; 32] = [
-    166, 230, 6, 120, 41, 228, 53, 167, 211, 20, 34, 171, 34, 191, 116, 23, 134,
-    105, 138, 238, 229, 146, 92, 206, 255, 57, 14, 164, 52, 159, 126, 13
-];
-// (rev) 0d7e9f34a40e39ffce5c92e5ee8a69861774bf22ab2214d3a735e4297806e6a6
-
 // Taproot test vectors from BIP-341 state the hashes without any reversing
-sha256t_hash_newtype!(TapLeafHash, TapLeafTag, MIDSTATE_TAPLEAF, 64,
-    doc="Taproot-tagged hash for elements tapscript Merkle tree leafs", forward
-);
-sha256t_hash_newtype!(TapBranchHash, TapBranchTag, MIDSTATE_TAPBRANCH, 64,
-    doc="Taproot-tagged hash for elements tapscript Merkle tree branches", forward
-);
-sha256t_hash_newtype!(TapTweakHash, TapTweakTag, MIDSTATE_TAPTWEAK, 64,
-    doc="Taproot-tagged hash for elements public key tweaks", forward
-);
-sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
-    doc="Taproot-tagged hash for the elements taproot signature hash", forward
-);
+sha256t_hash_newtype! {
+    pub struct TapLeafTag = hash_str("TapLeaf/elements");
+    /// Taproot-tagged hash for elements tapscript Merkle tree leafs".
+    #[hash_newtype(forward)]
+    pub struct TapLeafHash(_);
+
+    pub struct TapBranchTag = hash_str("TapBranch/elements");
+    /// Tagged hash used in taproot trees; see BIP-340 for tagging rules", forwar
+    #[hash_newtype(forward)]
+    pub struct TapNodeHash(_);
```

Comment:
Should be `TapBranchHash` ?

Reply:
Upstream we renamed this to `TopNodeHash` in https://github.com/rust-bitcoin/rust-bitcoin/pull/1479 in order to simplify the single-leaf Taproot case.